File: FYS-MEK1110-mechanics/NewtonsCradle.m
function NewtonsCradle()
% Newtons Cradle - Mandatory assignment nr.7 FYS-MEK1110

N = 3;
m = 1.0;
k =1000.0;
q = 1.0;
d = 0.1;
v0 = 1.0;
time = 1.0;
dt = 0.001;

n = ceil(time/dt);
x = zeros(n,N);
v = zeros(n,N);
F = zeros(n,N);
t = zeros(n,1);

for j=1:N
    x(1,j) = d*(j-1);
end
v(1,1)=v0;
for i=1:n-1
    % Find the forces oon each block,j.
    % First, force from block to the left.
    for j=2:N
        dx = x(i,j)-x(i,j-1);
        F(i,j)=F(i,j)+force(dx,d,k,q);
    end
    % Second, force from block to the right
    for j=1:N-1
        dx = x(i,j+1)-x(i,j);
        F(i,j) = F(i,j) - force(dx,d,k,q);
    end
    % Euler-Chromer:
    
    a = F(i,:)/m;
    v(i+1,:) = v(i,:) + a*dt;
    x(i+1,:) = x(i,:) + v(i+1,:)*dt;
    t(i+1,1) = t(i,1) + dt;
end



    function F = force(dx,d,k,q)
        if dx<d
            F = k*abs(dx-d).^q;
        else
            F = 0.0;
        end
    end

% Plotting the results:

fontsize=18;
legends = ['ball 1', 'ball 2', 'ball 3'];
f = figure()
for j=1:N
    plot(t,v(:,j));
    tmp = ['Ball ' num2str(j,'%g')];
    if j==1
        hold on;
    end
    if j==N
        hold off
    end
end

set(gca,'FontSize',fontsize)
%get(0,'DefaultAxesColorOrder')
name = ['Newtons Cradle. k=' num2str(k,'%.1f') ' q=' num2str(q,'%.1f')];
title(name)
xlabel('t [s]')
ylabel('v,[m/s]')
legend('ball 1','ball 2', 'ball 3')
Title = ['plot_NewtonsCradle_k' num2str(k,'%g') '_q' num2str(q,'%g') '.png'];
print(f,'-dpng', Title) % save figure as png with name 'name'

end

File: FYS-MEK1110-mechanics/NewtonsCradle.py
# Newtons Cradle simulates a Newtons cradle.

import numpy as np
from scitools.std import *

def force(dx,d,k,q):
    if dx<d:
        F = k*abs(dx-d)**q
    else:
        F = 0.0
    return F

N = 3      # Number of balls
m = 1.0     # mass of balls
k = 100.0    # [N/m] 
#q = 4.0      # exponent
#q = 3.0/4   # exponent 
q = 1.0     # exponent
d = 0.1     # [m] distance between balls
v0 = 1.1    # [m/s] initial velocity of ball A
time = 1.0    # [s] simulation time
dt = 0.001  # [s] timestep
n = int(round(time/dt)) # number of timesteps
x = zeros((n,N),float) # position of every ball (0,1,...,N) at time ti = 0,1,2,...,n
v = x.copy()
t = zeros(n,float)
##########################################################
# Initial conditions

v[0,0] = v0       # initial velocity of ball 0

for i in range(N):
    x[0,1] = d*i # intial position of ball 0,1,...,N

for i in range(n-1):                         # Timeloop
    # Find force in vector F
    F = zeros(N,float)                       # update forces
    for j in range(1,N):                     # loop over balls 1:N = 1,2...,N
        #print 'loop range(1,N), j=%g' %j
        dx = x[i,j] - x[i,j-1]
        F[j] = F[j] + force(dx,d,k,q)
    for j in range(N-1):                     # loop over 
        dx = x[i,j+1] - x[i,j]
        F[j] = F[j] - force(dx,d,k,q)

    # Euler-Cromer vectorized step:

    a = F/m
    v[i+1] = v[i] + a*dt
    x[i+1] = x[i] + v[i+1]*dt
    t[i+1] = t[i] + dt
    

#########################################################
# plotting:

legends = []
figure()
for j in range(N):
    plot(t,v[:,j])
    temp = str(j)
    legends.append('Ball '+ temp)
    if j==0:
        hold('on')
    if j==N-1:
        hold('off')

title('Velocity. Number of balls N=%g' % N)
xlabel('time [s]')
ylabel('velocity [m/s]')
legend((legends), loc='center right')
#hardcopy('cradleOfNewton_ex_h',fontsize=25) # sets the fontsize in PostScript points only

legends = []
figure()
for j in range(N):
    plot(t,x[:,j])
    temp = str(j)
    legends.append('Ball '+ temp)
    if j==0:
        hold('on')
    if j==N-1:
        hold('off')

title('Position of balls. Number of balls N=%g' % N)
xlabel('time [s]')
ylabel('position [m]')
legend((legends), loc='center right')
#hardcopy('cradleOfNewton_ex_h',fontsize=25) # sets the fontsize in PostScript points only


print 'v/v0 = ' , v[n-1,:]/v0
